### PR TITLE
[Chore] Refactoring the EnStateHandler.

### PR DESCRIPTION
**Pull request by haosap**
## Proposal
The current way to use the EnStateHandler is little chaos. See the graph below:

### Before
image

### After
image

In general, the ENStateHandler receives the states from ReachbilityService and ExposureManager. But the current implementation gets the exposure state from HomeInteractor. Meanwhile, the EnStateHandler is injected everywhere. In fact, the injected objects are only interested in the State of EnStateHandler instead of the EnStateHandler itself. In addition, we follow the single source of truth policy. Only the SceneDelegate is responsible for instantiate the EnStateHandler and propagates the changing events to other objects who implements the following protocol:
```swift
 protocol ENStateHandlerUpdating: AnyObject {
	func updateEnState(_ state: ENStateHandler.State)
} 
```

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/SceneDelegate.swift [hunk 1]
```diff
@@ -133,22 +157,41 @@ final class SceneDelegate: UIResponder, UIWindowSceneDelegate, RequiresAppDepend
...
+				//TODO: Set some state
 				self?.presentHomeVC()
 			}
 		}
 	}
 
 	private func presentHomeVC() {
+
+		//FIXME: Move it to other place.
+		let enStateHandler = ENStateHandler(
```
Comment (steins-code) [inline src/xcode/ENA/ENA/Source/SceneDelegate.swift:189]: current enStateHandler needs to be accessible from SceneDelegate ... Proposal: extract it out of the method

File: src/xcode/ENA/ENA/Source/Scenes/Home/HomeInteractor.swift [hunk 2]
```diff
@@ -77,6 +73,7 @@ final class HomeInteractor {
 	private var exposureSubmissionService: ExposureSubmissionService?
 	var stateHandler: ENStateHandler!
```
Comment (steins-code) [inline src/xcode/ENA/ENA/Source/Scenes/Home/HomeInteractor.swift:74]: remove ENStateHandler it will be nil with that approach

File: src/xcode/ENA/ENA/Source/Scenes/ENSetting/ExposureNotificationSettingViewController.swift [hunk 3]
```diff
@@ -231,3 +228,12 @@ private extension ENSettingModel.Content {
 		}
 	}
 }
+
+// MARK: ENStateHandler Updating
+extension ExposureNotificationSettingViewController: ENStateHandlerUpdating {
+	func updateEnState(_ state: ENStateHandler.State) {
```
Comment (steins-code) [inline src/xcode/ENA/ENA/Source/Scenes/ENSetting/ExposureNotificationSettingViewController.swift:234]: This method is not called when the exposurenotification settings was opened via SettingsviewController and thus the status is wrong in this case

File: src/xcode/ENA/ENA/Source/Models/Home/HomeActivateCellConfigurator.swift [hunk 4]
```diff
@@ -60,12 +60,14 @@ final class HomeActivateCellConfigurator: CollectionViewCellConfigurator {
...
-
 	func setupAccessibility(for cell: ActivateCollectionViewCell) {
 		cell.isAccessibilityElement = true
 		cell.accessibilityIdentifier = Accessibility.StaticText.homeActivateTitle
 	}
 }
+
+extension HomeActivateCellConfigurator: ENStateHandlerUpdating {
+	func updateEnState(_ state: ENStateHandler.State) {
+		self.state = state
```
Comment (steins-code) [inline src/xcode/ENA/ENA/Source/Models/Home/HomeActivateCellConfigurator.swift:71]: When the internet is turned off this card isnt being updated without scrolling in Homescreen

Comment (steins-code) [reply src/xcode/ENA/ENA/Source/Models/Home/HomeActivateCellConfigurator.swift:71]: Same for turning on internet again

File: src/xcode/ENA/ENA/Source/Models/Exposure/__tests__/Mocks/MockStateHandlerObserverDelegate.swift [hunk 5]
```diff
@@ -19,30 +19,9 @@
 import ExposureNotification
 import XCTest
```
Comment (steins-code) [inline src/xcode/ENA/ENA/Source/Models/Exposure/__tests__/Mocks/MockStateHandlerObserverDelegate.swift:20]: All Regression tests are complaining :) please fix them

File: src/xcode/ENA/ENA/Source/Models/Home/HomeActivateCellConfigurator.swift (hunk 4 above)
Comment (haosap) [reply src/xcode/ENA/ENA/Source/Models/Home/HomeActivateCellConfigurator.swift:71]: fixed.

File: src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift [hunk 6]
```diff
@@ -69,20 +86,25 @@ class ENStateHandler {
 				currentState = .internetOff
 			}
 		case .internetOff:
-			guard let latestState = delegate?.getLatestExposureManagerState() else {
-				return
-			}
-			currentState = determineCurrentState(from: latestState)
+			//FIXME: What does this mean?
```
Comment (steins-code) [inline src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift:89]: Remove unneeded code

File: src/xcode/ENA/ENA/Source/Models/Exposure/__tests__/Mocks/MockStateHandlerObserverDelegate.swift (hunk 5 above)
Comment (haosap) [reply src/xcode/ENA/ENA/Source/Models/Exposure/__tests__/Mocks/MockStateHandlerObserverDelegate.swift:20]: fixed.

File: src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift [hunk 7]
```diff
@@ -17,38 +17,55 @@
...
-	case restricted
-}
 
-protocol StateHandlerObserverDelegate: AnyObject {
-	func stateDidChange(to state: RiskDetectionState)
-	func getLatestExposureManagerState() -> ExposureManagerState
+
+protocol ENStateHandlerUpdating: AnyObject {
+	func updateEnState(_ state: ENStateHandler.State)
+	//var initialState: ENStateHandler.State
```
Comment (steins-code) [inline src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift:24]: remove unneeded line

File: src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift [hunk 8]
```diff
@@ -17,38 +17,55 @@
...
+		case enabled
+		/// Exposure Notification is disabled.
+		case disabled
+		/// Bluetooth is off.
+		case bluetoothOff
+		/// Internet is off.
+		case internetOff
+		/// Restricted Mode.
+		case restricted
+		//FIXME: NOT YET DONE.
```
Comment (steins-code) [inline src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift:41]: will this be fixed in different PR?

File: src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift [hunk 9]
```diff
@@ -17,38 +17,55 @@
 
 import Foundation
```
Comment (steins-code) [inline src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift:19]: satisfy swiftlint :)

**Review by steins-code — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift (hunk 8 above)
Comment (haosap) [reply src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift:41]: yes. Next one about ENManager. :-)

File: src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift (hunk 7 above)
Comment (haosap) [reply src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift:24]: fixed.

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 10]
```diff
@@ -129,12 +129,12 @@ final class ENAExposureManager: NSObject, ExposureManager {
 
 		self.observer = observer
 
-		enabledObservation = observe(\.manager.exposureNotificationEnabled, options: .new) { [weak self] _, _ in
-			guard let self = self else { return }
-			DispatchQueue.main.async {
-				observer.exposureManager(self, didChangeState: self.preconditions())
-			}
-		}
+//		enabledObservation = observe(\.manager.exposureNotificationEnabled, options: .new) { [weak self] _, _ in
```
Comment (ChristianKienle) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:132]: ?

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 11]
```diff
@@ -108,7 +108,7 @@ final class ENAExposureManager: NSObject, ExposureManager {
 	// MARK: Properties
 
 	private weak var observer: ENAExposureManagerObserver?
-	private var enabledObservation: NSKeyValueObservation?
+//	private var enabledObservation: NSKeyValueObservation?
```
Comment (ChristianKienle) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:111]: ?

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 12]
```diff
@@ -174,6 +174,12 @@ final class ENAExposureManager: NSObject, ExposureManager {
 	}
 
 	private func changeEnabled(to status: Bool, completion: @escaping CompletionHandler) {
+//		print("The status of exposure manager is \(manager.exposureNotificationStatus)")
```
Comment (ChristianKienle) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:177]: ?

File: src/xcode/ENA/ENA/Source/Models/Home/HomeActivateCellConfigurator.swift [hunk 13]
```diff
@@ -21,9 +21,9 @@ final class HomeActivateCellConfigurator: CollectionViewCellConfigurator {
 
 	let identifier = UUID()
 	
-	private var state: RiskDetectionState
+	private var state: ENStateHandler.State
 
-	init(state: RiskDetectionState) {
+	init(state: ENStateHandler.State) {
```
Comment (ChristianKienle) [inline src/xcode/ENA/ENA/Source/Models/Home/HomeActivateCellConfigurator.swift:26]: `EN`-prefix questionable

File: src/xcode/ENA/ENA/Source/SceneDelegate.swift [hunk 14]
```diff
@@ -92,6 +92,10 @@ final class SceneDelegate: UIResponder, UIWindowSceneDelegate, RequiresAppDepend
 		return HTTPClient(configuration: config)
 	}()
 
+
+	private var enStateHandler:ENStateHandler?
```
Comment (ChristianKienle) [inline src/xcode/ENA/ENA/Source/SceneDelegate.swift:96]: Space missing

File: src/xcode/ENA/ENA/Source/SceneDelegate.swift [hunk 15]
```diff
@@ -146,22 +174,44 @@ final class SceneDelegate: UIResponder, UIWindowSceneDelegate, RequiresAppDepend
 	}
 
 	private func showHome(animated _: Bool = false) {
+		//FIXME: During the onboarding, if the user decline, the status == Unknown.
+		//After that
 		if exposureManager.preconditions().status == .active {
 			presentHomeVC()
 		} else {
-			log(message: "ExposureManager not activate yet.")
+//			let enManager = ENManager()
```
Comment (ChristianKienle) [inline src/xcode/ENA/ENA/Source/SceneDelegate.swift:182]: ?

File: src/xcode/ENA/ENA/Source/SceneDelegate.swift [hunk 16]
```diff
@@ -146,22 +174,44 @@ final class SceneDelegate: UIResponder, UIWindowSceneDelegate, RequiresAppDepend
...
 
 	private func presentHomeVC() {
+
+		enStateHandler = ENStateHandler(
+				initialExposureManagerState: exposureManager.preconditions(),
+				reachabilityService: ConnectivityReachabilityService(),
+				delegate: self)
+
+
+		guard let enStateHandler = self.enStateHandler else {
```
Comment (ChristianKienle) [inline src/xcode/ENA/ENA/Source/SceneDelegate.swift:211]: can be removed

File: src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift [hunk 17]
```diff
@@ -17,38 +17,52 @@
...
+	private weak var delegate: ENStateHandlerUpdating?
 	private var internetOff = false
+	private var latestExposureManagerState: ExposureManagerState
 
 	init(
-		_ initialState: ExposureManagerState,
-		reachabilityService: ReachabilityService,
-		delegate: StateHandlerObserverDelegate
-	) {
+			initialExposureManagerState:ExposureManagerState,
```
Comment (ChristianKienle) [inline src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift:58]: space missing

File: src/xcode/ENA/ENA/Source/Scenes/ENSetting/ExposureNotificationSettingViewController.swift [hunk 18]
```diff
@@ -64,7 +57,18 @@ final class ExposureNotificationSettingViewController: UITableViewController {
 		navigationItem.largeTitleDisplayMode = .always
 		setUIText()
 		tableView.sectionFooterHeight = 0.0
+
 	}
+//
+//	private func tryEnManager() {
```
Comment (ChristianKienle) [inline src/xcode/ENA/ENA/Source/Scenes/ENSetting/ExposureNotificationSettingViewController.swift:63]: ?

File: src/xcode/ENA/ENA/Source/Scenes/Home/HomeInteractor.swift [hunk 19]
```diff
@@ -534,18 +530,11 @@ extension HomeInteractor {
 	}
 }
 
-extension HomeInteractor: StateHandlerObserverDelegate {
-	func stateDidChange(to _: RiskDetectionState) {
+//MARK: - The ENStateHandler updating
+extension HomeInteractor: ENStateHandlerUpdating {
+	func updateEnState(_ state: ENStateHandler.State) {
+		self.enState = state
```
Comment (ChristianKienle) [inline src/xcode/ENA/ENA/Source/Scenes/Home/HomeInteractor.swift:536]:
```suggestion
		enState = state
```

**Review by ChristianKienle — APPROVED** [review]
super
